Develop

KonashinAlexander opened the pull request:
Sprint-12

KonashinAlexander committed:
build(package): add redux

KonashinAlexander committed:
build(package): add logger

KonashinAlexander committed:
feat(store): add store, reducer

KonashinAlexander committed:
components

KonashinAlexander committed:
feat(burger-constructor): add store, add total price

KonashinAlexander committed:
feat(store): add currIngredient add order

KonashinAlexander committed:
feat(store): add ordeId

KonashinAlexander committed:
feat(order): add POST request

KonashinAlexander committed:
feat(app): add counter to buns

KonashinAlexander committed:
feat(constructor): add sort

KonashinAlexander committed:
fix(ingredients): change counter

KonashinAlexander committed:
fix(constr-element): trash btn

KonashinAlexander committed:
fix(constr elem): change design

KonashinAlexander committed:
feat(readme): add changes

gennady-bars commented on src/utils/api.js:
```diff
@@ -7,12 +9,15 @@ export const checkResponse = (res) => {
-            .catch(err=>{console.log(err)})
-}
\ No newline at end of file
+            .catch(err=>{console.log('getIngredients error >>', err)})
```
Не стоит использовать `catch` в `Api`, так как ловить ошибки нужно в самом конце цепочки запроса, а конец цепочки находится после всего  кода, который происходит в `then`. И в реальных проектах в `catch` происходит много логики: открывается попап с ошибкой и тд. Поэтому файл `api` не может знать, что делать с ошибкой внутри.

KonashinAlexander replied:
catch()  закомментировал пока.
надо сказать, что несколько раз данный catch() срабатывал,

gennady-bars replied:
> надо сказать, что несколько раз данный catch() срабатывал,

он-то и срабатывает, но дальше по коду идет продолжение цепочки, которое имеет тоже код, где могут быть ошибки

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -1,86 +1,125 @@
+    const getOrder = (ingredientsIds) => {
+        const newOrder = {"ingredients": ingredientsIds}
+
+        return fetch(`${API_URL}/orders`, {
```
Все **асинхронные** функции проекта нужно сделать в виде `thunk`-функций редакса и объявить их в папке `actions`

Вы используете `createAsyncThunk`, значит, и отправка заказа тоже будет с `createAsyncThunk` там в `order.js`

KonashinAlexander replied:
Поправил, сделал по аналогии с ingredients.js

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -1,86 +1,125 @@
+              'Content-type': 'application/json; charset=UTF-8',
+            },
+          })
+            .then(checkResponse)
```
**Можно лучше**

Можно сделать **универсальную** функцию запроса с **проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

И даже можно поместить внутрь `baseUrl`, чтобы не дублировать его в каждом запросе. Тогда нужно будет передавать просто `endpoint` в вызов

KonashinAlexander replied:
Попробовал, работает.

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -1,86 +1,125 @@
+          })
+            .then(checkResponse)
+            .then(dataOrder => {
+                if(dataOrder.success) {
```
`.success`

проверка на успех везде повторяется


**Можно лучше**


Если интересно, то можно создать **универсальную** функцию запроса `request` , которая будет содержать в себе уже **базовый урл**, проверку на **ok** и проверку на `success`, чтобы не дублировать их в каждом запросе.

```js

// 1 раз объявляем базовый урл
export const BASE_URL = "https://norma.nomoreparties.space/api/";

// создаем функцию проверки ответа на `ok`
const checkResponse = (res) => {
  if (res.ok) {
    return res.json();
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ошибка ${res.status}`);
};

// создаем функцию проверки на `success`
const checkSuccess = (res) => {
  if (res && res.success) {
    return res;
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ответ не success: ${res}`);
};

// создаем универсальную фукнцию запроса с проверкой ответа и `success`
// В вызов приходит `endpoint`(часть урла, которая идет после базового) и опции
const request = (endpoint, options) => {
  // а также в ней базовый урл сразу прописывается, чтобы не дублировать в каждом запросе
  return fetch(`${BASE_URL}${endpoint}`, options)
    .then(checkResponse)
    .then(checkSuccess);
};

```


И теперь не нужно будет дублировать все это в каждом запросе. Нужно просто заменить в запросах все `fetch` на `request`, удалив лишние проверки и базовый урл. Все остальное останется без изменений

Код станет намного чище. Будет мелкий аналог `axios`

Пример запроса ингредиентов

```js
const getIngredients =  () => request("ingredients");

```

В `get`-запросах даже не нужно указывать 2й параметр. Он получится `undefined`, что без проблем для `fetch` (который внутри).

KonashinAlexander replied:
Попробовал, работает.

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -1,86 +1,125 @@
-    const otherIngredients = useMemo(()=>constructorIngredients.filter(data => data.name !== 'Краторная булка N-200i'),[constructorIngredients]);
+      const renderOtherIngredients = useCallback((data, index) => {
+        return (
+            <ConstructorElementItem {...data} key={data.uuid} index={index} id={index}/>
```
`                key={data.uuid} 
`

Отлично, что указываете **уникальный** `uuid` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`

KonashinAlexander replied:
Спасибо!

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -1,86 +1,125 @@
+                </div>   
+                       
+                <ol className={cn(style.box_big, 'mt-4')} ref={drop} >
+                    {otherIngredients.map((data, i) => renderOtherIngredients(data, i))}     
```
`                    {otherIngredients.map((data, i) => renderOtherIngredients(data, i))}     
`

Если внутри **безымянной** (стрелочной) функции вызывается одна функция с **точно такими же** аргументами, то эта безымянная функция не нужна. Это лишняя обертка вокруг самой функции, которую можно использовать просто сразу в коде


```jsx
                    {otherIngredients.map(renderOtherIngredients)}     



```

KonashinAlexander replied:
Поправил, работает.

gennady-bars commented on src/components/burger-ingredient/burger-ingredient.jsx:
```diff
@@ -1,7 +1,49 @@
+            <p className="text text_type_main-default">{props.name}</p>            
+        </div> 
+
+        {showModal && <Modal title='Детали ингредиента' onClose={closeModal}>
```
Любой `JSX` нужно оборачивать в `круглые скобки`, чтобы избежать проблем с автоматической вставкой точек с запятой - такое иногда бывает при компиляции


```jsx

      {showModal && ( <========= Круглые скобочки =========
        <Modal title="Детали ингредиента" onClose={closeModal}>
          <IngredientDetails data={showModal} />
        </Modal>
      )} <========= Круглые скобочки =========

```



Это можно сделать комбинацией клавиш `SHIFT + ALT + F` или с помощью `prettier`

gennady-bars replied:
**Можно лучше**

 лучше не делать в **каждом** ингредиенте по модальному окну. Нужно было сделать одно окно для деталей в ` BurgerIngredients` (или `App`) и передавать в каждый ингредиент функцию **клика**, в которой бы назначались данные для показа в окне. Это поможет в спринте, где **роутинг** будете изучать и реализовывать

KonashinAlexander replied:
Стиль поправил.
Модальные окна пока не стал трогать, необходимо некоторое время чтобы разобраться.

gennady-bars commented on src/components/constructor-element/constructor-element.jsx:
```diff
@@ -0,0 +1,81 @@
+import { useDrag, useDrop } from 'react-dnd';
+
+
+const ConstructorElementItem = ({ id, index, image, name, price, ...props}) => {
```
`ConstructorElementItem` забыли типизировать

gennady-bars replied:
Нужно посмотреть все компоненты и **типизировать** их `пропсы`, если где-то пропустили.

KonashinAlexander replied:
Исправил!

gennady-bars commented on src/components/ingredient-details/ingredient-details.jsx:
```diff
@@ -24,4 +25,21 @@ export const IngredientDetails = ({data}) => {
+
+IngredientDetails.propTypes = {
+    data: PropTypes.shape({
+        __v: PropTypes.number,
+        _id: PropTypes.string,
+        calories: PropTypes.number,
+        carbohydrates: PropTypes.number,
+        fat: PropTypes.number,
+        image: string,
+        image_large: PropTypes.string,
+        image_mobile: PropTypes.string,
+        name: PropTypes.string,
+        price: PropTypes.number,
+        proteins: PropTypes.number,
+        type: PropTypes.string
```
у Вас есть тип ингредиента `ingredientsPropType`  в `utils`


Нужно его везде использовать. чтобы не дублировать


Просто его нужно сразу с `PropTypes.shape` указать


```js

export const ingredientsPropType = PropTypes.shape({
  __v: PropTypes.number,
  _id: PropTypes.string,
  calories: PropTypes.number,
  carbohydrates: PropTypes.number,
  fat: PropTypes.number,
  image: string,
  image_large: PropTypes.string,
  image_mobile: PropTypes.string,
  name: PropTypes.string,
  price: PropTypes.number,
  proteins: PropTypes.number,
  type: PropTypes.string,
});

```

KonashinAlexander replied:
Исправил

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -5,15 +5,20 @@ import cn from 'classnames';
         function closeOnEsc(e) {
           if (e.key === "Escape" || e.key === "Esc") {
             onClose();
+            dispatch(removeCurrentIngredient())
```
`            dispatch(removeCurrentIngredient())
`

это нужно удалить

`Modal` является **универсальным** компонентом для любых модальных окон. Он не должен знать внутри себя какие-то **конкретные** данные (экшены) для конкретных попапов.

KonashinAlexander replied:
Поправил.

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -29,7 +34,10 @@ export const Modal = ({ title, onClose, children }) => {
-                <CloseIcon onClick={onClose}/>
+                <CloseIcon onClick={()=>{
+                                        onClose()
+                                        dispatch(removeCurrentIngredient())
```
`            dispatch(removeCurrentIngredient())
`

это нужно удалить

`Modal` является **универсальным** компонентом для любых модальных окон. Он не должен знать внутри себя какие-то **конкретные** данные (экшены) для конкретных попапов.

KonashinAlexander replied:
Поправил.

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично, что указываете **уникальный** `uuid` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`
* Хорошая логика `DnD` и редакса


но есть некоторые недочеты: 

* Скрин https://disk.yandex.ru/i/FxEEYIwgkP0LJg лишние 3 файла нужно удалить 
* `error-boundary.js` нужно переложить в `components`. Это компонент реакта
* В консоли предупреждения https://disk.yandex.ru/i/6H_k1p0fxLEikw 
* Не стоит использовать `catch` в `Api`, так как ловить ошибки нужно в самом конце цепочки запроса, а конец цепочки находится после всего  кода, который происходит в `then`. И в реальных проектах в `catch` происходит много логики: открывается попап с ошибкой и тд. Поэтому файл `api` не может знать, что делать с ошибкой внутри.
* Все **асинхронные** функции проекта нужно сделать в виде `thunk`-функций редакса и объявить их в папке `actions`
* Любой `JSX` нужно оборачивать в `круглые скобки`, чтобы избежать проблем с автоматической вставкой точек с запятой
* Вам нужно **отформатировать** код, чтобы были видны нормальные отступы и вложенность одних компонентов в другие. И `js`-код тоже будет удобно читать. Это можно сделать комбинацией клавиш `SHIFT + ALT + F` или с помощью `prettier`
* `ConstructorElementItem` забыли типизировать
* Нужно посмотреть все компоненты и **типизировать** их `пропсы`, если где-то пропустили. 
* `Modal` является **универсальным** компонентом для любых модальных окон. Он не должен знать внутри себя какие-то **конкретные** данные (экшены) для конкретных попапов.






**Можно лучше**

* Желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
* лучше не давать нажать на кнопку `Оформить заказ`, пока пользователь не добавил булку в заказ.
* лучше отображать **лоадер** (загрузку) при ожидании ответа от сервера с номером заказа


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

KonashinAlexander committed:
fix(all) add PropTypes checking, remove App

KonashinAlexander committed:
fix(order): add thunk

KonashinAlexander committed:
fix(api): add universal func

KonashinAlexander committed:
fix(ingredient): change modal
  method: 'POST',
  body: JSON.stringify(newOrder),
  headers: {
    'Content-type': 'application/json; charset=UTF-8',
  }
}enter the commit message for your changes. Lines starting

gennady-bars commented on src/components/modal/modal.jsx:
```diff
@@ -1,56 +1,52 @@
-                }}/>
-            </div>
-            {children}
+      <div className={style.modal}>
```
```js

    top: 230px;
    left: 20%;
```


Модальное окно должно быть всегда ровно посередине экрана. Для этого нужно указать 
```css


  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);

```

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
